### PR TITLE
Remove duplicate CiliumNode watcher

### PR DESCRIPTION
**Pull request by aanm**
In certain cases the Cilium Operator could end up with two CiliumNode
k8s watchers. This PR re-writes the code in a way that a single
watcher will be used for all cases.

### CODE REVIEW

File: operator/cilium_node.go
Comment (jrajahalme) [inline operator/cilium_node.go]: I would think that a `....Builder` is a function type. Maybe better name this simply as `ciliumNodeWatcherConfig`??

File: operator/cilium_node.go
Comment (jrajahalme) [inline operator/cilium_node.go]: Same comment, please drop the `Builder` suffix :-)

File: operator/cilium_node.go
Comment (jrajahalme) [inline operator/cilium_node.go]: There is a lot of code duplication here, would it be better to only have one set of resource event handler funcs that would work both with nil and non-nil nodeManager?

File: operator/cilium_node.go
Comment (jrajahalme) [inline operator/cilium_node.go]: Do we really want 3 sets of these funcs with repeated conversion and error handling?

**Review by jrajahalme — APPROVED** [review]
Looks OK on the top level, but I was not happy with 3 sets of handler funcs which all repeat the same conversion and error handling.

File: operator/cilium_node.go
Comment (aanm) [reply operator/cilium_node.go]: @jrajahalme how would we distinguish between this struct and the other one from line 43?

**Review by joestringer — APPROVED** [review]
LGTM from docs perspective. Looks like Jarno already reviewed the core functionality here, so I skipped that for now. Let me know if you would like additional review.

File: operator/cilium_node.go
Comment (aanm) [reply operator/cilium_node.go]: Same comment above.

File: operator/cilium_node.go
Comment (aanm) [reply operator/cilium_node.go]: There is, I was worried on optimizing the code over readbility. Please see the commit _operator: simplify duplicated code in the event handling_ after this one where I've simplified the code a little bit.

**Comment by aanm** [comment]
@joestringer @ldelossa since this code changed so much in this PR I decided to go ahead and added in the last commit to have the workqueue pattern that we have discussed a couple weeks ago. Let me know what you think. /cc @christarazi FYI

File: operator/cilium_node.go
Comment (joestringer) [inline operator/cilium_node.go]: Build builds what? :joy: 
```suggestion
// Build builds the configuration and if WithKVStoreSync was called it will start a SharedStore.
```

File: operator/main.go
Comment (joestringer) [inline operator/main.go]: nit: It seems error-prone to me to initialize this object partially here, partially a hundred lines down, partially another hundred lines down, then finish initialization and start goroutines etc just below that. Can we just track a few booleans and initialize & start it all at once?

More specifically, I think it's a concerning sign if the builder pattern (ie `foo.WithBar()`) is used all by itself and not as part of a block that does:
```
foo :=NewFoo().WithAlice()
if bar {
    foo = foo.WithBar()
}
if bob {
    foo = foo.WithBob()
}
```
The reasoning is similar; basically it's hard for the reader to understand that the code 100 lines down can safely modify the structure in a way that an initializer function like this would do. How do we know that some other logic isn't already consuming the structure and we're just trashing memory others are accessing? Well, we can dig around and find all the references to the structure.... or in the first place we just initialize once in a single block, then it's easier to tell if it's doing the right thing.

File: operator/main.go
Comment (joestringer) [inline operator/main.go:410]: If the nodewatcher is no longer run above this block, could we miss handling of some events correctly?

File: operator/main.go
Comment (joestringer) [inline operator/main.go]: Is there no context for the operator? :thinking: we should add one (maybe just file an issue, could be a good-first issue maybe?).

File: operator/cilium_node.go
Comment (ldelossa) [reply operator/cilium_node.go]: I don't think it's unreasonable to have a structure called "...Builder" with a method set that builds something. In this case, a config.

File: operator/cilium_node.go
Comment (ldelossa) [inline operator/cilium_node.go]: You can probably drop the atomic.Value for a simple integer type and call atomic.AddInt32/atomic.LoadInt32. The API maybe a bit nicer to deal with.

File: operator/cilium_node.go
Comment (ldelossa) [inline operator/cilium_node.go]:
```suggestion
	// We shouldn't modify cfgBuilder.nodeManager after a
```

File: operator/main.go
Comment (joestringer) [inline operator/main.go]: I'm confused why the builder needs to exist here, why couldn't we just do something like:
```suggestion
	cnSync, err := NewCiliumNodeSynchronizer(cnWatcherCfg)
	if err != nil {
		log.WithError(err).Fatal("Unable to generate configuration for CiliumNode watcher")
	}
	if err := cnSync.startSynchronizingCiliumNodes(); err != nil {
```

File: operator/cilium_node.go
Comment (ldelossa) [inline operator/cilium_node.go]:
```suggestion
// Build builds the config and if WithKVStoreSync was called it will start a SharedStore.
```

File: operator/cilium_node.go
Comment (joestringer) [inline operator/cilium_node.go]: What does this comment mean? I looked up https://pkg.go.dev/k8s.io/client-go/util/workqueue#RateLimitingInterface and I'm now more confused which conditions or methods this comment is trying to explain.

File: operator/cilium_node.go
Comment (joestringer) [inline operator/cilium_node.go]: I feel like this code is trying to trick me. At face value, the actual execution for each item that's processed is "do nothing", but actually the "do something" is a byproduct of the iterator. This code would be easier to follow if it were just opencoding the implementation like `for key, quit := cfg.queue.Get(); !quit { doStuff() }`.

File: operator/cilium_node.go
Comment (ldelossa) [inline operator/cilium_node.go]: I think for readability take, pull out these function literals to somewhere outside the if/else block and give them semantic meaning. Such as "AddFuncNilNodeManager"

File: operator/cilium_node.go
Comment (joestringer) [inline operator/cilium_node.go:224]: When does this quit? Is there any possibility that this will somehow stop processing events but the overall operator will continue running just without the functionality that's supposed to be handled in this workqueue?

(Maybe put another way, when this quits, should we log an Info log or something that the operator will stop processing these work items?)

File: operator/cilium_node.go
Comment (ldelossa) [inline operator/cilium_node.go:121]: Do we want to handle those context.TODO()'s before merging?

**Review by ldelossa — CHANGES_REQUESTED** [review]

**Review by joestringer — COMMENTED** [review]
Some minor comments from a deeper review (mostly focused on the first and last commits) but I think the implementation looks correct.

Mainly I found the Builder structure confusing. I'm used to reading Cilium code that says something like, 

```
func NewFooer(cfg Config) *fooer { return &fooer{....}
func (f *fooer) WithBar(b Bar) *fooer { f.Bar = b; return f }
func (f *fooer) Run() {
    if f.FooEnabled {
        doFoo()
    }
    if f.BarEnabled {
        doBar()
    }
}
```

However, here there's no Fooer. There's a config builder (which does stuff like syncing state with the kvstore, which is a bit surprising), then that returns a config which is passed to a static function that operates against that config.

I'm sure there's reasons for this approach, and maybe some of it is that k8s code looks more like this so I'm just out of my depth a bit. But I at least found it difficult to follow.

File: operator/cilium_node.go
Comment (joestringer) [reply operator/cilium_node.go]: (On second read, I see that the actual execution is running `processNextWorkItem()` so it's clear that it will do something, but it's the empty braces that mislead me here.)

File: operator/main.go
Comment (aanm) [reply operator/main.go:410]: The are independent, the nodewatcher is not used in the `if kvstoreEnabled() {` branch.

Comment (joestringer) [reply operator/main.go:410]: Looking more deeply into the `runNodeWatcher()` function, it just kicks off two async goroutines to GC CNP/CCNP node status anyway so the previous code wasn't providing any guarantees about when it would run. So should be fine from that perspective.

File: operator/cilium_node.go
Comment (aanm) [reply operator/cilium_node.go:224]: This should only quit when the operator terminates.

**Review by aanm — COMMENTED** [review]
Thanks for the review

**Review by ldelossa — APPROVED** [review]
My feedback was handled. Approving contingent on other's reviews.

File: operator/cilium_node.go
Comment (joestringer) [inline operator/cilium_node.go]: nit: I suspect that these were added in response to my earlier feedback, but a couple of these comments are superfluous, the code `queue.Get()` clearly gets the next event from the queue. Similarly the `queue.Done()` call marks it as done, that's clear from the code. If that's how the workqueue interface works, we don't have a choice and the comments don't help explain why, they just duplicate the explanation from the code in different words. I think it's fine to drop those comments.

```suggestion
	key, quit := queue.Get()
	if quit {
		return false
	}
```

**Review by joestringer — APPROVED** [review]
LGTM. I'm sure the more I look, the more I could find ways we can incrementally tidy up the code but I think it'll just become bikeshedding at that point. :ship:

**Comment by aanm** [comment]
/test-1.22-4.9

**Review by jrajahalme — APPROVED** [review]
LGTM

File: operator/cilium_node.go
Comment (christarazi) [inline operator/cilium_node.go]: I recall @ldelossa mentioning some nuance where in some case (forgot what it was) that we need to make some extra call to `queue.Add()` or similar. It that addressed here? Louis, could you remind me what that was and validate whether this code already addressed it?

**Review by christarazi — COMMENTED** [review]
Changes LGTM, one question below regarding workqueue handling.

File: operator/cilium_node.go
Comment (aanm) [reply operator/cilium_node.go]: Nice call @christarazi I've added that missing piece.

**Comment by aanm** [comment]
I'll merge this PR since there was only one line added and the previous CI run was green IIRC.:
```
--- a/operator/cilium_node.go
+++ b/operator/cilium_node.go
@@ -232,9 +232,6 @@ func processNextWorkItem(queue workqueue.RateLimitingInterface, syncHandler func
                return true
        }
 
+       log.WithError(err).Errorf("sync %q failed with %v", key, err)
+       queue.AddRateLimited(key)
+
        return true
 }
 ```
 
 This line will only be executed on error cases which is unlikely to be caught by our CI.

**Pull request merged**
